CommonCode/include/EffCorrFactor.h

```diff
@@ -78,33 +86,161 @@ void EffCorrFactor::write(TFile& OutputFile, string effArgName, TH2D* _h_num, TH
+         exit(1);
+      }
+
+      for (Int_t iy=0;iy<=h_2D_bfCorr->GetNbinsY();iy++)
+      {
+         // checking the bin boundaries 
+         if (h_2D_bfCorr->GetYaxis()->GetBinLowEdge(iy)!=_heff_2D->GetYaxis()->GetBinLowEdge(iy) ||
+             h_2D_bfCorr->GetYaxis()->GetBinLowEdge(iy)!=h_2D_afCorr->GetYaxis()->GetBinLowEdge(iy))
+         {
+            printf("[Error] EffCorrFactor::applyEffCorrOnHisto is doing a matrix multiplication.\n" \
+                   "This requires the uncorrected, corrected distributions and the efficiency factor has the same bin configuration!\n");
+            exit(1);
+         }
+
+         double x = h_2D_bfCorr->GetXaxis()->GetBinCenter(ix);
+         double y = h_2D_bfCorr->GetYaxis()->GetBinCenter(iy);
+
+         int hBinIdx    = h_2D_bfCorr->FindBin(x,y);
+         int effBinIdx  = _heff_2D->FindBin(x,y);
+         if(_heff_2D->GetBinContent(effBinIdx)>0)
+         {
+            double n_afCorr   = h_2D_bfCorr->GetBinContent(hBinIdx)/_heff_2D->GetBinContent(effBinIdx);
+            double errrel_num = (h_2D_bfCorr->GetBinError(hBinIdx)>0 && h_2D_bfCorr->GetBinContent(hBinIdx)>0)? h_2D_bfCorr->GetBinError(hBinIdx)/h_2D_bfCorr->GetBinContent(hBinIdx): 0;
+            double errrel_den = _heff_2D->GetBinError(effBinIdx)/_heff_2D->GetBinContent(effBinIdx);
+            
+            double errrel_n_afCorr = TMath::Sqrt(errrel_num*errrel_num+errrel_den*errrel_den);
+
+            h_2D_afCorr->SetBinContent(hBinIdx,n_afCorr);
+            h_2D_afCorr->SetBinError  (hBinIdx,n_afCorr*errrel_n_afCorr);
+         }
+         else
+         {
+            h_2D_afCorr->SetBinContent(hBinIdx,0);
+            h_2D_afCorr->SetBinError(hBinIdx,0);
+         }
+         // if (ix<=2 and iy <=2)
+         // {
+         //    printf("[ix=%d,x=%.3f][iy=%d,y=%.3f] h_2D_bfCorr=%.3f+/-%.3f, _heff_2D=%.3f+/-%.3f, h_2D_afCorr=%.3f+/-%.3f\n", 
+         //             ix, x, iy, y,
+         //             h_2D_bfCorr->GetBinContent(hBinIdx), h_2D_bfCorr->GetBinError(hBinIdx),
+         //             _heff_2D->GetBinContent(effBinIdx), _heff_2D->GetBinError(effBinIdx),
+         //             h_2D_afCorr->GetBinContent(hBinIdx), h_2D_afCorr->GetBinError(hBinIdx));
+         // }
+      }
+   }
+}
+
 Float_t EffCorrFactor::efficiency(Float_t argBin)
 {
-  Float_t e = _heff_1D->GetBinContent(_heff_1D->FindBin(argBin));
-  if(e < 0.00000000001){
-    if(counter < 10){
-      std::cout << "!!!Error on efficiency correction! Zero efficiency!!! " << _heff_1D->GetName() << "=" << argBin << std::endl << std::endl;
-      if(counter == 9) std::cout << " !!!Greater than ten calls of this error... TERMINATING OUTPUT, PLEASE FIX" << std::endl;
-      ++counter;
-    }
-    
+   if (!_heff_1D) 
+   {
+      printf("[Error] EffCorrFactor::applyEffCorrOnHisto couldn't find _heff_1D.\n");
+      _effInf->ls();
+      exit(1);
```

Comment:
This is generally not advised to exit the program like this, especially in header files.  It’s better to print an alert message and return some dummy value.